### PR TITLE
Update entities in batch when field encryption settings change

### PR DESCRIPTION
**Pull request by svendecabooter**
This PR fixes the following:
- Re-enable the functionality to update fields when the encryption settings get changed, that was disabled in the last PR (https://github.com/d8-contrib-modules/field_encrypt/pull/25)
- Implement Batch API to update the entities in groups of 5 after field storage settings are changed.

This has been tested with the following scenario's:
- Non-encrypted field gets encryption --> plain text data gets encrypted
- Encrypted field (with method X) settings get changed to use method Y --> encrypted data gets re-encrypted with the new encryption method
- Encrypted field made unencrypted (checkbox unchecked) --> encrypted data gets stored as plain text

### CODE REVIEW

**Comment by svendecabooter** [comment]
After the update is performed, old values still seem to be in the cache.
Can't quite figure out how to remove them from the cache... normally the node:[NID] cache invalidation gets called on node save, so they should be removed...

**Comment by nerdstein** [comment]
@svendecabooter i am really curious if the Batch API is the way to go. Doesn't this signify a lapse of time between the encryption settings of a field (updated) and the potential data updates needed to process the change? Maybe this is resolved by revisions or something, but I need to understand this a lot more from your point of view

**Comment by svendecabooter** [comment]
I'm not sure I understand the concern exactly.
You mean there would be a problem if entities get created in the meantime?

Upon submitting the form to change the field settings, we query the database to get all entities with this field, and store the original (pre-update) encryption settings.

The batch is then run on each of the entities, (optionally) decrypting with the old settings, and (re)encrypting with the newly saved settings.

While the batch is running, the new settings are already saved, so new entities would be encrypted with the correct settings.

**Comment by svendecabooter** [comment]
I guess there would be a problem if entities get updated manually, while the batch is running...

I think realistically we only have 2 choices:
- Work like core: once there is data in the field, the storage settings can no longer be updated. This is the easiest to preserve data integratity, but not very user friendly.
- Allow the option to change the settings (and perform encryption / decryption even when data exists). Though this should probably be marked with a big warning about data integrity. 
  It could be very helpful if you want to encrypt data that was already stored in plaintext earlier, but doing it in a live and changing environment can indeed cause problems, so a decent backup and a controlled environment to perform the updates seem like necessities.

**Comment by svendecabooter** [comment]
Either way this feature is a pretty big game of whack-a-mole, with all the possible variations (multiple languages, multiple revisions, multiple bundles) and encryption / decryption actions that get performed.

Every time I fix a use case, another use case breaks again, so this will need thorough testing, if we can get it into a state where every use case seems to work.

**Comment by nerdstein** [comment]
@svendecabooter understood and I'm comfortable making incremental progress. Let's hop on a hangout and chat.

**Comment by svendecabooter** [comment]
The latest commits make the code work for 3 scenario's mentioned above (start encrypting existing field, change encryption method on existing field, stop encrypting existing field), tested with the following use cases:
- regular entity
- entity with revisions
- entity with multiple translations

I can start writing some tests to confirm this, if we agree to add this feature.

**Comment by nerdstein** [comment]
@svendecabooter I gave this a lot more thought and yes the Batch API is probably the most elegant way to get this done. I was confusing myself a bit thinking through all of this, but nothing that wasn't cured with a good night's sleep.

I do think, if possible, we could "lock" records when we process them inside of the batch API. Do you know of any features?

I am not talking about "code" level locking, as demonstrated here: https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Lock%21LockBackendInterface.php/group/lock/8

I believe each entity storage has a "locked" attribute (as seen here: https://api.drupal.org/api/drupal/core!modules!field!src!Entity!FieldStorageConfig.php/class/FieldStorageConfig/8).

Maybe it's possible for us to set `locked=true` for the entire field before processing the batch API records. Or, maybe this is overkill altogether. I would love to hear your thoughts.

Regardless, I want to stress we're on the right track with this. I'm going to review the code in depth and provide any inline comments right now

File: field_encrypt.module [hunk 1]
```diff
@@ -34,6 +33,11 @@ function field_encrypt_form_alter(&$form, FormStateInterface $form_state, $form_
         '#open' => TRUE,
       );
 
+      // Display a warning about changing field data.
+      if ($form_id == "field_storage_config_edit_form" && $field->hasData()) {
+        $form['field_encrypt']['#prefix'] = '<div class="messages messages--warning">' .t('Warning: changing field encryption settings may cause data corruption!<br />When changing these settings, existing fields will be (re)encrypted in batch according to the new settings. <br />Make sure you have a proper backup, and do not perform this action in an environment where the data will be changing during the batch operation, to avoid data loss.') . '</div>';
```
Comment (nerdstein) [inline field_encrypt.module]: Awesome

File: field_encrypt.module [hunk 2]
```diff
@@ -100,11 +105,12 @@ function field_encrypt_form_alter(&$form, FormStateInterface $form_state, $form_
  */
 function field_encrypt_form_field_add_form_builder($entity_type, \Drupal\field\Entity\FieldStorageConfig $field_storage_config, &$form, \Drupal\Core\Form\FormStateInterface $form_state) {
   $field_encryption_settings = $form_state->getValue('field_encrypt');
+  $field_encryption_settings['encrypt'] = (bool) $field_encryption_settings['encrypt'];
   $form_encryption = $field_encryption_settings['encrypt'];
   $original_encryption = $field_storage_config->getThirdPartySettings('field_encrypt');
 
   // If the form has the value, we set it.
-  if ($form_encryption === 1) {
+  if ($form_encryption) {
```
Comment (nerdstein) [inline field_encrypt.module]: Nit picks, this variable seems excessive - shouldn't this just be `$field_encryption_settings['encrypt']`?

File: field_encrypt.module [hunk 3]
```diff
@@ -117,33 +123,123 @@ function field_encrypt_form_field_add_form_builder($entity_type, \Drupal\field\E
...
+/**
+ * Submit handler to update encryption after saving field storage settings.
+ *
+ * @param $form
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ */
+function field_encrypt_update_encryption_form_submit(&$form, FormStateInterface $form_state) {
+  $storage = $form_state->getStorage();
+  $entity_type_manager = \Drupal::service('entity_type.manager');
+  if (isset($storage['field_encrypt_update']) && $storage['field_encrypt_update'] == TRUE) {
```
Comment (nerdstein) [inline field_encrypt.module:153]: Can you add some more inline comments here? Especially around batch API operations

File: field_encrypt.module [hunk 4]
```diff
@@ -117,33 +123,123 @@ function field_encrypt_form_field_add_form_builder($entity_type, \Drupal\field\E
...
+    $entity_ids = $query->execute();
+
+    $batch = array(
+      'title' => t('Updating field encryption'),
+      'operations' => array(
+        array('field_encrypt_update_field_encryption', array(array_keys($entity_ids), $field_name, $entity_type_id, $original_encryption_settings)),
+      ),
+      'finished' => 'field_encrypt_update_field_encryption_finished',
+    );
+    batch_set($batch);
```
Comment (nerdstein) [inline field_encrypt.module:186]: cool deal

File: field_encrypt.module [hunk 5]
```diff
@@ -117,33 +123,123 @@ function field_encrypt_form_field_add_form_builder($entity_type, \Drupal\field\E
...
+//  $entity_type_manager = \Drupal::service('entity_type.manager');
+  $process_entities = \Drupal::service('field_encrypt.process_entities');
+  if (empty($context['sandbox'])) {
+    $context['sandbox']['entity_ids'] = $entity_ids;
+    $context['sandbox']['progress'] = 0;
+    $context['sandbox']['max'] = count($entity_ids);
+  }
+
+  // Process entities by groups of 5.
+  $count = min(5, count($context['sandbox']['entity_ids']));
```
Comment (nerdstein) [inline field_encrypt.module]: Do you think we can/should make this configurable? Maybe in the settings form, we allow the user to select how many get processed at once? Default it to 5?

File: field_encrypt.module [hunk 6]
```diff
@@ -117,33 +123,123 @@ function field_encrypt_form_field_add_form_builder($entity_type, \Drupal\field\E
...
+ * @param array $results
+ *   The value set in $context['results'] by callback_batch_operation().
+ * @param array $operations
+ *   If $success is FALSE, contains the operations that remained unprocessed.
+ */
+function field_encrypt_update_field_encryption_finished($success, $results, $operations) {
+  // The 'success' parameter means no fatal PHP errors were detected. All
+  // other error management should be handled using 'results'.
+  if ($success) {
+    $message = \Drupal::translation()->formatPlural(count($results), 'One field updated.', '@count fields updated.');
```
Comment (nerdstein) [inline field_encrypt.module:245]: Is it possible to say which field we updated in the message, like "Body field updated"?

File: field_encrypt.module [hunk 7]
```diff
@@ -117,33 +123,123 @@ function field_encrypt_form_field_add_form_builder($entity_type, \Drupal\field\E
...
+  }
+  else {
+    $message = t('Finished with an error.');
```
Comment (nerdstein) [inline field_encrypt.module:248]: Is there any way we can get more context? This is so generic... Will there be any exceptions thrown? Or, error messages reported elsewhere?

**Comment by nerdstein** [comment]
Code reviewed and it looks really good. Minor comments.

One other potential black hole I thought of:

Are we assuming field config changes happen only on production? These settings are in CMI …. would we possibly need the batch API to run if these config changes happen on another environment? E.g. someone updates a YML file w/ field encrypt settings that are then pushed up to another environment

**Comment by nerdstein** [comment]
Black hole comments:

https://pantheon.io/blog/hard-things-are-possible-configuration-management-drupal-8

Config API has some mediation events through event subscribers (it seems). An example: https://api.drupal.org/api/drupal/core!lib!Drupal!Core!Config!ConfigEvents.php/constant/ConfigEvents%3A%3ASAVE/8

Full set of events listed here: https://api.drupal.org/api/drupal/core!lib!Drupal!Core!Config!ConfigEvents.php/class/ConfigEvents/8

In short, it seems we need an EventSubscriber that monitors config changes to FieldSettings. If we see encrypted == true and a change between existing field encryption settings, we invoke our same batch API code (in theory and maybe not in practice).

File: field_encrypt.module (hunk 6 above)
Comment (svendecabooter) [inline field_encrypt.module:245]: I've been looking at this, but that isn't immediately available in the variables at that point. Maybe there's a way to do this, but I'm going to focus my time on other issues first.

File: field_encrypt.module (hunk 7 above)
Comment (svendecabooter) [inline field_encrypt.module:248]: That should be possible, as long as we gather errors during the update process.
Those could then be captured and displayed in the batch results.
See Drupal\config\Form\ConfigSync::finishBatch for an example.

Will check if I can implement this still within this PR, or in a followup patch.

**Comment by svendecabooter** [comment]
I think this is probably good to be merged.

Followup issues for the remaining outstanding tasks:
- https://github.com/d8-contrib-modules/field_encrypt/issues/11
- https://github.com/d8-contrib-modules/field_encrypt/issues/22
- https://github.com/d8-contrib-modules/field_encrypt/issues/27

Let me know if I missed something

**Comment by nerdstein** [comment]
Merging
